Generalize grow, refactor math, random

pentacular opened the pull request:
Grow/Trim

This mostly changes grow to do a minkowski sum with a convex hull upon points, segments, polygons, and meshes.

Trim is added as a lower dimensional counterpoint -- given a mesh it grows the individual faces, given a polygon it grows the individual edges.

This allows, e.g., x.cut(trim(Orb(1))) to trim x by 1 mm in all dimenions.

The tool can be two dimensional, e.g., x.grow(Arc(1)) to grow in x and y by one mm without affecting z.

The tool can be asymmetric, e.g., x.grow(Box(1)) to grow in x and y by one mm, with sharp axis-aligned edges.

Or x.grow(Orb(cut(XY()))) to grow rounded in all directions other than positive z.

Random
The main use of random numbers so far is to produce a set of N values.

x.random(4, 12, (a, b, c, d) => ...) generates the first four numbers of seed 12 and binds them to a, b, c, d.

The Nth set of M random numbers can be selected with the third integer argument.

e.g., Seq({ to: 20 }, t => random(4, 12, t, (a, b, c, d) => ...))

will repeat 20 times using different groups of 4 random values.

BarbourSmith commented:
The tool can be asymmetric, e.g., x.grow(Box(1)) to grow in x and y by one mm, with sharp axis-aligned edges.

Or x.grow(Orb(cut(XY()))) to grow rounded in all directions other than positive z

This is slick

BarbourSmith commented on algorithm/cgal/Grow.h:
```diff
@@ -1,96 +1,226 @@
-  typedef CGAL::Exact_predicates_exact_constructions_kernel EK;
-  typedef CGAL::Surface_mesh<EK::Point_3> Surface_mesh;
-  size_t size = geometry->size();
+static int Grow(Geometry* geometry, size_t count) {
```
What does count do here?

pentacular replied:
Count is the number of geometries in the input (as opposed to the tool).

There's probably a better naming choice for that.

BarbourSmith commented on algorithm/cgal/computeBoundingBox.js:
```diff
@@ -19,7 +19,7 @@ export const computeBoundingBox = (inputs) => {
         return;
       default:
-        throw new Error(`Unexpected status ${status}`);
+        throw new Error(`Unexpected status ${status} in computeBoundingBox`);
```
These are nice 👍

BarbourSmith reviewed (approved):
Is it intensional that these bounding boxes all went away?

<img width="589" alt="image" src="https://github.com/jsxcad/JSxCAD/assets/9359447/4a85bf06-da5a-4fdd-be1c-7e3c185c08ec">

Does this look right?

<img width="1284" alt="image" src="https://github.com/jsxcad/JSxCAD/assets/9359447/e0ad4de0-5bad-4eb5-95d2-1c39fdbd9ce0">

pentacular commented:
> Is it intensional that these bounding boxes all went away?

Yes, I finally deleted all of the existing ones to force generation, since the bounding boxes were sometimes under the pixel threshold.

I'm hoping that gauges provide a better alternative.

> 
> <img alt="image" width="589" src="https://private-user-images.githubusercontent.com/9359447/329333555-4a85bf06-da5a-4fdd-be1c-7e3c185c08ec.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTUyOTk2MDEsIm5iZiI6MTcxNTI5OTMwMSwicGF0aCI6Ii85MzU5NDQ3LzMyOTMzMzU1NS00YTg1YmYwNi1kYTVhLTRmZGQtYmUxYy03ZTNjMTg1YzA4ZWMucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDUxMCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA1MTBUMDAwMTQxWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MmQwMWFlNzgwZGNmZTkyMmQ1OGYxOTcwYzQ5NGE0MmJlYzJlODhhNjg1YmM5ZWQ2NjBmMGYxOGNlNThlM2ZhNSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.jOy_eCpsDFUZBrHx2IGDI7BfyqOk3eLUVAO1EB_-GkI">
> Does this look right?
> 
> <img alt="image" width="1284" src="https://private-user-images.githubusercontent.com/9359447/329333999-e0ad4de0-5bad-4eb5-95d2-1c39fdbd9ce0.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTUyOTk2MDEsIm5iZiI6MTcxNTI5OTMwMSwicGF0aCI6Ii85MzU5NDQ3LzMyOTMzMzk5OS1lMGFkNGRlMC01YmFkLTRlYjUtOTVkMi0xYzM5ZmRiZDljZTAucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDUxMCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA1MTBUMDAwMTQxWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZTZmMTRmNmE4M2I0YmVmNWIwMmY2YjgzOTgyZmNjZDRiNDkyZTRmZWQ1MzA5MjYzYTdjZTUzNmQ2NTBkNzQ2YSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.s4jdAYKa3ldL5bi7-xSqTJMnkITMyoJrfl9jaVk9wb8">

Ah, thanks for noticing that.

I've removed fromPolygons and fixed fromPolygonSoup to make sure it bounds a volume when closed.
